types/marked/index.d.ts

```diff
@@ -122,86 +113,82 @@ declare namespace marked {
-        table(src: string): Tokens.Table;
-        lheading(src: string): Tokens.Heading;
-        paragraph(src: string): Tokens.Paragraph;
-        text(src: string): Tokens.Text;
-        escape(src: string): Tokens.Escape;
-        tag(src: string, inLink: boolean, inRawBlock: boolean): Tokens.Tag;
-        link(src: string): Tokens.Image | Tokens.Link;
-        reflink(src: string, links: Tokens.Link[] | Tokens.Image[]): Tokens.Link | Tokens.Image | Tokens.Text;
-        emStrong(src: string, maskedSrc: string, prevChar: string): Tokens.Em | Tokens.Strong;
-        codespan(src: string): Tokens.Codespan;
-        br(src: string): Tokens.Br;
-        del(src: string): Tokens.Del;
-        autolink(src: string, mangle: (cap: string) => string): Tokens.Link;
-        url(src: string, mangle: (cap: string) => string): Tokens.Link;
-        inlineText(src: string, inRawBlock: boolean, smartypants: (cap: string) => string): Tokens.Text;
+        space(src: string): NonNullable<Tokens.Space | T>;
+        code(src: string): NonNullable<Tokens.Code | T>;
+        fences(src: string): NonNullable<Tokens.Code | T>;
+        heading(src: string): NonNullable<Tokens.Heading | T>;
+        nptable(src: string): NonNullable<Tokens.Table | T>;
+        hr(src: string): NonNullable<Tokens.Hr | T>;
+        blockquote(src: string): NonNullable<Tokens.Blockquote | T>;
+        list(src: string): NonNullable<Tokens.List | T>;
+        html(src: string): NonNullable<Tokens.HTML | T>;
+        def(src: string): NonNullable<Tokens.Def | T>;
+        table(src: string): NonNullable<Tokens.Table | T>;
+        lheading(src: string): NonNullable<Tokens.Heading | T>;
+        paragraph(src: string): NonNullable<Tokens.Paragraph | T>;
+        text(src: string): NonNullable<Tokens.Text | T>;
+        escape(src: string): NonNullable<Tokens.Escape | T>;
+        tag(src: string, inLink: boolean, inRawBlock: boolean): NonNullable<Tokens.Tag | T>;
+        link(src: string): NonNullable<Tokens.Image | Tokens.Link | T>;
+        reflink(
+            src: string,
+            links: Tokens.Link[] | Tokens.Image[],
+        ): NonNullable<Tokens.Link | Tokens.Image | Tokens.Text | T>;
+        emStrong(src: string, maskedSrc: string, prevChar: string): NonNullable<Tokens.Em | Tokens.Strong | T>;
+        codespan(src: string): NonNullable<Tokens.Codespan | T>;
+        br(src: string): NonNullable<Tokens.Br | T>;
+        del(src: string): NonNullable<Tokens.Del | T>;
+        autolink(src: string, mangle: (cap: string) => string): NonNullable<Tokens.Link | T>;
+        url(src: string, mangle: (cap: string) => string): NonNullable<Tokens.Link | T>;
+        inlineText(
+            src: string,
+            inRawBlock: boolean,
+            smartypants: (cap: string) => string,
+        ): NonNullable<Tokens.Text | T>;
     }
 
-    class Renderer {
+    type TokenizerObject = Partial<Omit<Tokenizer<false>, "constructor" | "options">>;
+
+    class Renderer<T = null> {
```

Comment:
I don't get it, what's the point of this type parameter?

Reply:
The Renderer function can only return a string when used as a Renderer class object but can return a string or `false` as a a Renderer extension object. (`false` means use the previous renderer) This parameter was the only way I know to allow the functions to return different things based on what type of object it was.

I could just copy all of the methods to a new type where the only difference is what is able to be returned but I figured it would be better this way than needing to keep two types in sync if a function is changed.

Reply:
Wouldn't it be better to use `never` for default type instead of `null`. I think you could then drop all of the `NonNullable` annotations.